Let the formatter format the whitespaces when converting an ArrowClauseExpression to a ReturnStatement when applicable

sdelarosbil opened the pull request:
Closes #71636 

The issue was that the expression part of the tree gets converted from the existing nodes which do not features any elastic trivia, but as the issue explains, these trivia may require the formatter's intervention to get right because the indentation should change after the conversion. This isn't always the case: it's possible indentations is already fine and can be preserved (there's 3 tests with this case already where the return is a LINQ fluent chain)

However, in cases where the arrow clause doesn't end on a newline, it means we know for sure the indentation has to change. I accomplish this by replacing all whitespaces with elastic empty ones which allows the formatter to figure it out.

Let me know if these criteria are appropriate and if the way I introduce the elastic trivia are too.

sharwell reviewed (changes requested):
In general, it is not desirable to use elastic trivia for reindentation purposes. This will cause the formatter to change the formatting even within the expression (whitespace which is not part of indentation), which breaks the rule that formatting outside the scope of a specific refactoring should be preserved where possible.

This refactoring should follow the same pattern used by Convert to File Scoped Namespace and Convert to Block Scoped Namespace, where lines within the change region are explicitly indented to account for the change.

If this is not possible and we decide to proceed with the sub-optimal strategy of overwriting the user's formatting, it can be done in a simpler fashion by just adding the formatter annotation to the resulting node that needs to be formatted.

sdelarosbil commented:
> In general, it is not desirable to use elastic trivia for reindentation purposes. This will cause the formatter to change the formatting even within the expression (whitespace which is not part of indentation), which breaks the rule that formatting outside the scope of a specific refactoring should be preserved where possible.
> 
> This refactoring should follow the same pattern used by Convert to File Scoped Namespace and Convert to Block Scoped Namespace, where lines within the change region are explicitly indented to account for the change.
> 
> If this is not possible and we decide to proceed with the sub-optimal strategy of overwriting the user's formatting, it can be done in a simpler fashion by just adding the formatter annotation to the resulting node that needs to be formatted.
> 
> @sdelarosbil Note that most of our features have not yet been updated to avoid extraneous formatting options, as this is a fairly recent addition to the strategy. One of my concerns in this PR is the fact that the new behavior is implemented in a shared component and not limited to a specific refactoring. Shared behavior is an ideal location to go with the new behavior. If the formatter annotation is used instead, it would ideally be limited to just the refactoring in question.

I partially addressed the concerns, let me explain.

I checked how the file scoped namespace refactor did it and unfortunately, it doesn't seem easy to use that strategy as is and it would imply refactoring I am not comfortable doing at this point. So I will opt in to do the aub optimal strategy.

I ended up moving the logic much further up. I originally thought more refactors actions were affected, but upon closer inspections, it seems that only the initialize parameters ones were concerned which allowed me to move the logic to only affect 3 refactors.

However, I am not sure what you mean by ` just adding the formatter annotation to the resulting node that needs to be formatted.` because I tried to add the elastic annotation to the expression of the return statement and it had no effect. I could only get my test to pass if I also messed with the descendant nodes.

sdelarosbil commented:
@sharwell Hey, this is a reminder as my previous comment is still valid as I need more details to properly address your review.

sharwell commented:
> because I tried to add the elastic annotation to the expression of the return statement and it had no effect

Formatter annotation is separate from elastic:
https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/Workspaces/Core/Portable/Formatting/Formatter.cs#L28-L31

sdelarosbil commented:
> > because I tried to add the elastic annotation to the expression of the return statement and it had no effect
> 
> Formatter annotation is separate from elastic:
> 
> https://github.com/dotnet/roslyn/blob/e7cae311ab49148335e969b94fb47f0dffaa95df/src/Workspaces/Core/Portable/Formatting/Formatter.cs#L28-L31

@sharwell  This formatter doesn't seem to work here when I add it to the expression of the return statement. Even if I aggressively add this annotation to all the descendant nodes, it has no effect. Only the elastic one arranges the formatting correctly.

Am I doing something wrong?

sdelarosbil commented:
@sharwell I applied your suggestion. If there's nothing else, I am fine to merge with squash.

sharwell commented:
@CyrusNajmabadi any issues with the approach here?

CyrusNajmabadi commented:
looking.

CyrusNajmabadi commented on src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs:
```diff
@@ -71,6 +73,15 @@ public static bool IsImplicitConversion(Compilation compilation, ITypeSymbol sou
+                    arrowClause.ArrowToken.TrailingTrivia.IndexOf(SyntaxKind.EndOfLineTrivia) == -1)
+                {
+                    var whiteSpaceTrivia = convertedReturn.Expression.DescendantTrivia().Where(static tr => tr.IsWhitespace() && !tr.IsElastic());
+                    convertedStatement = convertedStatement.ReplaceTrivia(whiteSpaceTrivia, (_, tr) => tr.AsElastic());
```
interesting idea.  we haven't really done this elsewhere.  but i have no issues with it.

CyrusNajmabadi commented on src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs:
```diff
@@ -71,6 +73,15 @@ public static bool IsImplicitConversion(Compilation compilation, ITypeSymbol sou
                     return;
                 }
 
+                if (convertedStatement is ReturnStatementSyntax convertedReturn &&
+                    convertedReturn.Expression is not null &&
```
```suggestion
                if (convertedStatement is ReturnStatementSyntax { Expression: not null } convertedReturn &&
```

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InitializeParameter/AddParameterCheckTests.cs:
```diff
@@ -1085,6 +1085,43 @@ private void Init()
+
+                class C
+                {
+                    int M(int v, [||]string s) => v switch
```
can you add a test with a conditional expression as well, of the forms:

```
 int M(bool v, [||]string s) => v ? 0 : 1;
 int M(bool v, [||]string s) => v ?
    0 :
    1;
 int M(bool v, [||]string s) => v
    ? 0
    : 1;
```

And variations where the expr starts on the next line.  And variations where the `=>` is on the next line indented?

sdelarosbil replied:
Done.

CyrusNajmabadi commented on src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs:
```diff
@@ -71,6 +73,15 @@ public static bool IsImplicitConversion(Compilation compilation, ITypeSymbol sou
+                if (convertedStatement is ReturnStatementSyntax convertedReturn &&
+                    convertedReturn.Expression is not null &&
+                    body is ArrowExpressionClauseSyntax arrowClause &&
+                    arrowClause.ArrowToken.TrailingTrivia.IndexOf(SyntaxKind.EndOfLineTrivia) == -1)
```
```suggestion
                    arrowClause.ArrowToken.TrailingTrivia.IndexOf(SyntaxKind.EndOfLineTrivia) < 0)
```

CyrusNajmabadi commented:
I'm ok with this with beefed up testing.  thanks!

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InitializeParameter/AddParameterCheckTests.cs:
```diff
@@ -1085,6 +1085,204 @@ private void Init()
+                            throw new ArgumentNullException(nameof(s));
+                        }
+                
+                        return v ? 0 : 1;
```
so this seems undesirable fwiw.

CyrusNajmabadi replied:
@sharwell thoughts on this outcome?  we could let this go, and treat it as an unfortunate consequence of this strategy.  or we could only take a pr here if it didn't undo the user's desired formatting in a case like this.

sharwell replied:
I believe this is just a side effect of the strategy used here. Once we implement a proper re-indenter that we can use in general refactorings, we could migrate from elastic trivia over to that new service and this issue would resolve itself.

CyrusNajmabadi commented on src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs:
```diff
@@ -71,6 +73,14 @@ public static bool IsImplicitConversion(Compilation compilation, ITypeSymbol sou
+                    arrowClause.ArrowToken.TrailingTrivia.IndexOf(SyntaxKind.EndOfLineTrivia) < 0)
+                {
+                    var whiteSpaceTrivia = convertedReturn.Expression.DescendantTrivia().Where(static tr => tr.IsWhitespace() && !tr.IsElastic());
+                    convertedStatement = convertedStatement.ReplaceTrivia(whiteSpaceTrivia, (_, tr) => tr.AsElastic());
```
ok.  based on the issues with how some of the tests changed, perhaps we coudl try just putting on elastic trivia on tokens that are at teh start of the line?

that said, i think that still won't necessarily work as my expectation then is that hte formatter would still undo things and place the items on the same line. 

What we genuinely want here is some sort of way of saying "fixup the indentation of these tokens, but otherwise do not touch them".  (or "fixup the indentation of this node, but otherwise don't touch it").

CyrusNajmabadi replied:
actually, since we have the function declaration, we can determinet he desired indentation.  we would then just want to take all leading tokens on every line and ensure it had that indentation if it was just preceded by whitespace on that line.  that seems simpl-ish enough.  importantly, it is targeted, and will likely not have any other downstream impacts.

sdelarosbil replied:
@CyrusNajmabadi I did some checking and first, I am not sure how we know the desired indentation by the function declaration because that's the unconverted version, not the converted one. I don't really see how I can get this information based on that.

But I did checked to see what would happen if I make only the line leading trivia elastic and it behaves unexpectedly, this is the test diff I got on TestUpdateExpressionBody4:

```
Xunit.Sdk.TrueException
Context: Code refactoring application
content of '/0/Test0.cs' did not match. Diff shown with expected as baseline:
 using System;
 
 class C
 {
     int M(int v, string s)
     {
         if (s is null)
         {
             throw new ArgumentNullException(nameof(s));
         }
 
         return v switch
-        {
-            _ => 0
-        };
+{
+    _ => 0
+    };
     }
 }
```
This is very strange. It only reformats correctly if I put every whitespaces elastic.

One thing I am a little perplexed on: the reason the return statement indents correctly is because it is preceded by a 0 width elastic, but if I try to use them here in place, now it removes all indent except the leading underscore one, but only of 4 spaces.

At this point, I am a bit confused if I should look into the formatter's code.

CyrusNajmabadi replied:
My guess is that this is a much larger and more complex item than could be done in a simple bug fix.  Specifically, i think our code-action pipeline needs the concept of "this node needs to be moved, and have its indentation changed, but otherwise have no otehr changes made to it".  We don't have such a concept right now, and it would need to be built out.

sdelarosbil replied:
Hmmm well in that case, I am not sure I can do that myself.

So, it's your call: if this method has too much side effects, feel free to close this. Should someone work on this, they'll have more to go on.
